core/src/main/java/eu/fasten/core/merge/CGMerger.java

```diff
@@ -386,6 +425,34 @@ public DirectedGraph mergeWithCHA(final DirectedGraph callGraph, final GraphMeta
         return result;
     }
 
+    /**
+     * Add a non resolved edge to the {@link DirectedGraph}.
+     */
+    private synchronized void addExternal(final MergedDirectedGraph result, final Set<LongLongPair> edges, Arc arc) {
+        for (String type : arc.target.receiverTypes) {
+            // Find external node URI
+            Map<String, String> typeMap = this.externalUris.get(type);
+            if (typeMap != null) {
+                String nodeURI = typeMap.get(arc.target.receiverSignature);
+
+                if (nodeURI != null) {
+                    // Find external node id
+                    Long target = this.allUris.inverse().get(nodeURI);
+                    if (target == null) {
+                        // Allocate a global id to the external node
+                        target = this.allUris.keySet().stream().max(Long::compareTo).orElse(0L) + 1;
```

Comment:
After the discussion, it seems that this line should be replaced with logic that assigns negative ids, as the "regular" ids are all positive. This allows us to have a simple "counter" with the next negative number and would allow us to get rid of the search through all keys.